File: 卡码/代码随想录/JS/字符串/05-KMP/LeetCode_0028.js
//28. 找出字符串中第一个匹配项的下标
const strStr = function(haystack, needle) {
  let next=new Array(needle.length);
  let j=0;
  next[j]=0;
  for(let i=1;i<needle.length;i++){
    while(j>0&&needle[i]!=needle[j]) j=next[j-1];
    if(needle[i]==needle[j]) j++;
    next[i]=j;
  }
  j=0;
  for(let i=0;i<haystack.length;i++){
    while(j>0&&needle[j]!=haystack[i]) j=next[j-1];
    if(needle[j]==haystack[i]) j++;
    if(j==needle.length) return i-j+1;
  }
  return -1;
};

File: 卡码/代码随想录/1/04-字符串/Go/05-KMP/LeetCode_0028.go
//28. 找出字符串中第一个匹配项的下标
package main

func strStr(haystack string, needle string) int {
	var next=make([]int,len(needle))
	j:=0
	next[j]=0
	for i:=1;i<len(needle);i++ {
		for j>0&&needle[j]!=needle[i] {
			j=next[j-1]
		}
		if needle[j]==needle[i] {
			j++
		}
		next[i]=j
	}
	j=0
	for i:=0;i<len(haystack);i++ {
		for j>0&&needle[j]!=haystack[i] {
			j=next[j-1]
		}
		if needle[j]==haystack[i] {
			j++
		}
		if j==len(needle) {
			return i-j+1
		}
	}
	return -1
}
